### PR TITLE
[SPARK-7270][SQL] Consider dynamic partition when inserting into hive table

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-7270

### CODE REVIEW

**Comment by viirya** [comment]
ping @yhuai @marmbrus.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 1]
```diff
@@ -760,7 +762,14 @@ private[hive] case class InsertIntoHiveTable(
   override def children: Seq[LogicalPlan] = child :: Nil
   override def output: Seq[Attribute] = child.output
 
-  override lazy val resolved: Boolean = childrenResolved && child.output.zip(table.output).forall {
+  val numDynamicPartitions = partition.values.count(_.isEmpty)
+  val tableOutput = {
```
Comment (marmbrus) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: If I understand correctly, this is the expected schema of the table that is generating the data to be inserted, possibly including dynamic partition columns.  Can you add a comment to that effect if I'm correct?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 2]
```diff
@@ -760,7 +762,14 @@ private[hive] case class InsertIntoHiveTable(
...
+    val hiveTable = table.asInstanceOf[MetastoreRelation]
```
Comment (marmbrus) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: This cast is worrisome.  Can we instead make `table` a `MetastoreRelation` so this is enforced when we create this operator?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 3]
```diff
@@ -760,7 +762,14 @@ private[hive] case class InsertIntoHiveTable(
...
+    (hiveTable.attributes ++ hiveTable.partitionKeys.takeRight(numDynamicPartitions))
+      .take(child.output.length)
```
Comment (marmbrus) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Why a take here?  I haven't worked through it, but it seems possible that we are going to silently ignore the case where the number of columns doesn't line up.

**Comment by marmbrus** [comment]
Thanks!  Merging to master and 1.4.

**Pull request closed without merging**